### PR TITLE
refactor: 13084 Split `Service` into `Service` and `RpcService` and migrated `Service` to platform-sdk

### LINKED ISSUE #13084: Refactor `Service` and move it to `platform-sdk`
The goal of this ticket is to make it possible for `PlatformState` to utilize schema framework that allows a service to have a protobuf definition and provides a way to migrate from one version to the other. 

To make it work, several things have to happen: 

1. Current service `Service` interface needs to be split in two parts:

`Service`: 
```
public interface Service {
    String getServiceName();
    void registerSchemas(@NonNull SchemaRegistry registry, @NonNull SemanticVersion version);
}
```
amd
`RpcService`:
```
public interface RpcService extends Service {   
    Set<RpcServiceDefinition> rpcDefinitions();
}
```
2. `RpcService` remains in `hedera-app-spi` while `Service` and a set of related classes such as:
```
MigrationContext
GenesisRecordsBuilder
NodeInfo
NetworkInfo
SelfNodeInfo
SchemaRegistry
StateDefinition
```
should migrate to `platform-sdk`. 

Here is how the class dependencies look like:

image

This refactoring will unblock the next refactoring https://github.com/hashgraph/hedera-services/issues/11771

### PR DESCRIPTION
**Pull request by imalygin**
**Description**:

Highlights of the PR: 
- Split `Service` interface into two interfaces - `RpcService` and `Service`. As the name suggests, `RpcService` comes with RPC definition. 
- Migrated `Service` interface and the classes it depends on to `swirlds-state-api`

**Related issue(s)**:

Fixes #13084 

**Notes for reviewer**:

This refactoring is required to make https://github.com/hashgraph/hedera-services/issues/11771 possible.

### CODE REVIEW

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/HapiUtils.java [hunk 1]
```diff
@@ -97,15 +97,6 @@ public class HapiUtils {
     public static final Comparator<Timestamp> TIMESTAMP_COMPARATOR =
             Comparator.comparingLong(Timestamp::seconds).thenComparingInt(Timestamp::nanos);
 
-    /** A {@link Comparator} for {@link SemanticVersion}s that ignores
```
Comment (imalygin) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/HapiUtils.java:100]: This code migrated to another `HapiUtils` class residing in `platform-sdk` - `com.swirlds.platform.state.spi.HapiUtils`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/spi/SchemaRegistry.java [hunk 2]
```diff
@@ -14,14 +14,13 @@
...
-package com.hedera.node.app.spi.state;
+package com.swirlds.platform.state.spi;
 
-import com.hedera.node.app.spi.Service;
 import edu.umd.cs.findbugs.annotations.NonNull;
 import java.util.Collection;
 
 /**
- * Provided by the application to the {@link Service}, the {@link SchemaRegistry} is used by the
+ * Provided by the application to the {@link }, the {@link SchemaRegistry} is used by the
```
Comment (mhess-swl) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/spi/SchemaRegistry.java:23]: `{@link }`?

File: hedera-node/hedera-file-service/src/main/java/module-info.java [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
     uses com.hedera.node.app.service.file.FileService;
 
+    requires com.swirlds.platform.core;
```
Comment (mhess-swl) [inline hedera-node/hedera-file-service/src/main/java/module-info.java:6]: Inversion of control typically would dictate that this API module should only have a dependency on another API module. Is that not possible here (or in other module-info files from this PR)?

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/services/ServiceRegistryImplTest.java
Comment (mhess-swl) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/services/ServiceRegistryImplTest.java:1]: Since this class is staying in the `hedera-app` module, should it be renamed to `RpcServiceRegistryImplTest` (along with `ServicesRegistryImpl`)?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/OrderedServiceMigrator.java [hunk 4]
```diff
@@ -40,7 +40,7 @@
  * The entire purpose of this class is to ensure that inter-service dependencies are respected between
  * migrations. The only required dependency right now is the {@link EntityIdService}, which is needed
  * for genesis blocklist accounts in the token service genesis migration. (See {@link
- * Service#registerSchemas(SchemaRegistry, SemanticVersion)}).
+ * RpcService#registerSchemas(SchemaRegistry, SemanticVersion)}).
```
Comment (mhess-swl) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/OrderedServiceMigrator.java:43]: `RpcService`, while an inheriting interface of `Service`, doesn't directly contain the `registerSchemas(...)` method. I think we want this to stay as `Service` (instead of `RpcService`)

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/spi/HapiUtils.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public final class HapiUtils {
+    /** A {@link Comparator} for {@link SemanticVersion}s that ignores
+     * any semver part that cannot be parsed as an integer. */
+    public static final Comparator<SemanticVersion> SEMANTIC_VERSION_COMPARATOR = Comparator.comparingInt(
+                    SemanticVersion::major)
+            .thenComparingInt(SemanticVersion::minor)
+            .thenComparingInt(SemanticVersion::patch)
+            .thenComparingInt(semVer -> HapiUtils.parsedIntOrZero(semVer.pre()))
+            .thenComparingInt(semVer -> HapiUtils.parsedIntOrZero(semVer.build()));
```
Comment (mhess-swl) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/spi/HapiUtils.java:34]: This PR makes it blatantly obvious that this comparator doesn't have explicit tests 😅

**Review by mhess-swl — COMMENTED** [review]
Aside from just a few little things, my general comment is that, since we have a hedera node SPI package, it seems appropriate to also have a hedera state SPI package instead of moving classes directly into `com.swirlds.platform`. Are there reasons for not doing this?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/grpc/impl/netty/NettyGrpcServerManager.java [hunk 6]
```diff
@@ -114,9 +115,14 @@ public NettyGrpcServerManager(
...
         requireNonNull(metrics);
 
-        // Convert the various RPC service definitions into transaction or query endpoints using the GrpcServiceBuilder.
+        // Convert the various RPC service definitions into transaction or query endpoints using the
+        // GrpcServiceBuilder.
         services = servicesRegistry.registrations().stream()
                 .map(ServicesRegistry.Registration::service)
+                // Not all services are RPC services, but here we need RPC services only. The main difference
+                // between RPC service and a service is that the RPC service has RPC definition.
+                .filter(v -> v instanceof RpcService)
```
[the remark this answers is not in the document]
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/grpc/impl/netty/NettyGrpcServerManager.java:124]: Could you please explain how it's better?

[the remark this answers is not in the document]
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/grpc/impl/netty/NettyGrpcServerManager.java:124]: >if v were of type (GrpcService extends RpcService), instanceof would fail. 

I don't think it is correct. `instanceof` operator (unlike `getClass()` equality) checks if the object's class is exactly the same as the provided class **or** the object's class has the provided class as parent. In this particular case the requirement for the object to implement `RpcService`. 

`isAssignableFrom(...)` expression may be beneficial, if we don't know the class we want against at compile time. Here it's not the case.

File: hedera-node/hedera-file-service/src/main/java/module-info.java (hunk 3 above)
Comment (imalygin) [reply hedera-node/hedera-file-service/src/main/java/module-info.java]: Fair enough. I moved the classes to `swirlds-state-api` module instead.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/services/ServiceRegistryImplTest.java
Comment (imalygin) [reply hedera-node/hedera-app/src/test/java/com/hedera/node/app/services/ServiceRegistryImplTest.java:1]: This class is still in `hedera-app` but so as `ServicesRegistry`. Also, `ServicesRegistry` is not limited to `RpcService`, so I don't think the renaming is required.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/OrderedServiceMigrator.java (hunk 4 above)
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/OrderedServiceMigrator.java]: Agreed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/spi/HapiUtils.java (hunk 5 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/spi/HapiUtils.java]: I added a comment there

**Comment by imalygin** [comment]
> Aside from just a few little things, my general comment is that, since we have a hedera node SPI package, it seems appropriate to also have a hedera state SPI package instead of moving classes directly into `com.swirlds.platform`. Are there reasons for not doing this?

I have to move these interfaces to `platform-sdk` to be able to refactor `PlatformState` to be a singleton node. It requires Schema and all infrastructure that is related to migration and representing the state as a PBJ object

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/spi/SchemaRegistry.java (hunk 2 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/spi/SchemaRegistry.java]: Fixed

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java [hunk 7]
```diff
@@ -57,6 +55,8 @@
 import com.swirlds.common.platform.NodeId;
 import com.swirlds.platform.listeners.ReconnectCompleteListener;
 import com.swirlds.platform.listeners.StateWriteToDiskCompleteListener;
+import com.swirlds.platform.state.spi.info.NetworkInfo;
+import com.swirlds.platform.state.spi.info.SelfNodeInfo;
```
[the remark this answers is not in the document]
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java]: Moved to `swirlds-state-api` instead

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java [hunk 8]
```diff
@@ -119,6 +118,7 @@
 import com.swirlds.platform.listeners.StateWriteToDiskCompleteListener;
 import com.swirlds.platform.state.PlatformState;
 import com.swirlds.platform.state.spi.WritableSingletonStateBase;
+import com.swirlds.platform.state.spi.worfklows.record.GenesisRecordsBuilder;
```
[the remark this answers is not in the document]
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java]: Moved to swirlds-state-api instead

File: hedera-node/hedera-app/src/itest/java/grpc/GrpcTestBase.java [hunk 9]
```diff
@@ -64,6 +66,7 @@
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import org.assertj.core.api.Assumptions;
+import org.jetbrains.annotations.NotNull;
```
[the remark this answers is not in the document]
Comment (imalygin) [reply hedera-node/hedera-app/src/itest/java/grpc/GrpcTestBase.java]: I removed all the usages of `org.jetbrains.annotations.NotNull`

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/grpc/impl/netty/NettyGrpcServerManager.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mhess-swl) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/grpc/impl/netty/NettyGrpcServerManager.java:124]: > or the object's class has the provided class as a parent
Huh, maybe this is a change in one of the more recent Java versions. 

I still think `isAssignableFrom(...)` is also correct, but there's no need to go on about it here :) thanks for checking!

**Comment by mhess-swl** [comment]
> > Aside from just a few little things, my general comment is that, since we have a hedera node SPI package, it seems appropriate to also have a hedera state SPI package instead of moving classes directly into `com.swirlds.platform`. Are there reasons for not doing this?
> 
> I have to move these interfaces to `platform-sdk` to be able to refactor `PlatformState` to be a singleton node. It requires Schema and all infrastructure that is related to migration and representing the state as a PBJ object

Per a discussion we had outside the PR, these classes will be moved into a separate, public API module in a different PR in order to keep the scope of this PR from getting even bigger.

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/HapiUtils.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+import com.hedera.hapi.node.base.SemanticVersion;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Comparator;
+
+/**
+ * Utility methods for the Hedera API.
+ */
+public final class HapiUtils {
+
+    // FUTURE WORK: Add unit tests for this class.
```
Comment (mhess-swl) [inline platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/HapiUtils.java:28]: Thanks!!

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mhess-swl) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java]: Much better place for it, thank you!

**Review by mhess-swl — APPROVED** [review]
LGTM! Tyvm @imalygin, this is a ton of work!

**Review by jeromy-cannon — APPROVED** [review]

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/state/NoOpGenesisRecordsBuilder.java [hunk 11]
```diff
@@ -33,7 +33,7 @@
  */
 
 import com.hedera.hapi.node.state.token.Account;
-import com.hedera.node.app.spi.workflows.record.GenesisRecordsBuilder;
+import com.swirlds.state.spi.worfklows.record.GenesisRecordsBuilder;
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/state/NoOpGenesisRecordsBuilder.java:36]: typo : workflows

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java [hunk 12]
```diff
@@ -96,6 +95,7 @@
 import com.swirlds.platform.system.status.PlatformStatus;
 import com.swirlds.platform.system.transaction.Transaction;
 import com.swirlds.state.HederaState;
+import com.swirlds.state.spi.worfklows.record.GenesisRecordsBuilder;
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:98]: Why should this be moved to `swirlds.state.spi` ?

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/info/NetworkInfo.java [hunk 13]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package com.hedera.node.app.spi.info;
+package com.swirlds.state.spi.info;
```
Comment (Neeharika-Sompalli) [inline platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/info/NetworkInfo.java:17]: This is not related to States API, Should this be moved? Are we moving everything under spi to `com.swirlds.state.` ?

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/info/NodeInfo.java [hunk 14]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package com.hedera.node.app.spi.info;
+package com.swirlds.state.spi.info;
```
Comment (Neeharika-Sompalli) [inline platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/info/NodeInfo.java:17]: Same here

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/info/SelfNodeInfo.java [hunk 15]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package com.hedera.node.app.spi.info;
+package com.swirlds.state.spi.info;
```
Comment (Neeharika-Sompalli) [inline platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/info/SelfNodeInfo.java:17]: same here

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/worfklows/record/GenesisRecordsBuilder.java [hunk 16]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package com.hedera.node.app.spi.workflows.record;
+package com.swirlds.state.spi.worfklows.record;
```
Comment (Neeharika-Sompalli) [inline platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/worfklows/record/GenesisRecordsBuilder.java:17]: same here

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/info/NetworkInfo.java (hunk 13 above)
Comment (imalygin) [reply platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/info/NetworkInfo.java:17]: I wish it wasn't related to States API but it is. Though indirectly 

image

In the initial implementation of this PR I decided to move this interface to `swirlds-platform-core` as currently contains the implementation details related to the state. 

There is a follow up ticket for this change: 
https://github.com/hashgraph/hedera-services/issues/13270

You intuition is correct, and `swirlds-state-api` is not the best location for these classes but it's still better than `swirlds-platform-sdk`. I didn't want to include this change in the current PR because it's already huge.

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/state/NoOpGenesisRecordsBuilder.java (hunk 11 above)
Comment (imalygin) [reply hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/state/NoOpGenesisRecordsBuilder.java]: Fixed

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/ServiceFactory.java [hunk 17]
```diff
@@ -22,7 +22,7 @@
 import java.util.ServiceLoader;
 
 /**
- * This class provides the ability to load {@link Service} implementations at runtime. The Java SPI
+ * This class provides the ability to load {@link RpcService} implementations at runtime. The Java SPI
```
Comment (artemananiev) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/ServiceFactory.java:25]: Should the class be renamed to RpcServiceFactory?

**Review by alittley — APPROVED** [review]

**Review by jeromy-cannon — DISMISSED** [review]

**Comment by jeromy-cannon** [comment]
@imalygin , instead of in the title `chore:`, perhaps `refactor:` would be a better fit.
[012: Commit Title Conventions](https://www.notion.so/swirldslabs/012-Commit-Title-Conventions-15b13dbb2a174a1b90cb19b4d07a399f?pvs=4)

**Comment by imalygin** [comment]
> @imalygin , instead of in the title `chore:`, perhaps `refactor:` would be a better fit. [012: Commit Title Conventions](https://www.notion.so/swirldslabs/012-Commit-Title-Conventions-15b13dbb2a174a1b90cb19b4d07a399f?pvs=4)



> @imalygin , instead of in the title `chore:`, perhaps `refactor:` would be a better fit. [012: Commit Title Conventions](https://www.notion.so/swirldslabs/012-Commit-Title-Conventions-15b13dbb2a174a1b90cb19b4d07a399f?pvs=4)

sure, that's a good idea, will do.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/ServiceFactory.java (hunk 17 above)
Comment (imalygin) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/ServiceFactory.java:25]: Yes, that's a good idea.

Comment (imalygin) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/ServiceFactory.java:25]: Done

**Review by mhess-swl — APPROVED** [review]

**Review by david-bakin-sl — APPROVED** [review]
LGTM w.r.t. smart contract service changes only - trivial necessary `import` statement modifications + one file that corrected two `@NotNull` annotations.

**Review by jeromy-cannon — APPROVED** [review]

**Review by artemananiev — APPROVED** [review]

**Review by tinker-michaelj — APPROVED** [review]
LGTM for now, tyvm @imalygin ...we can look to update Services to keep application-only infrastructure needed for migration in the `sharedValues` map.
